src/connection.rs

```diff
@@ -514,12 +402,66 @@ impl<T: AsyncRead + AsyncWrite + Unpin> Connection<T> {
 
                     return Err(ConnectionError::Closed);
                 }
-                Either::Right((frame, _)) => {
-                    if let Some(stream) = self.on_frame(frame.transpose().map_err(Into::into))? {
-                        return Ok(stream);
+                Event::NewStream(stream) => {
+                    return Ok(stream);
+                }
+            }
+        }
+    }
+
+    fn poll(&mut self, cx: &mut Context<'_>) -> Poll<Result<Event>> {
+        loop {
+            self.garbage_collect();
+
+            if self.socket.poll_ready_unpin(cx).is_ready() {
+                if let Some(frame) = self.pending_frames.pop_front() {
+                    self.socket.start_send_unpin(frame)?;
+                    continue;
+                }
+            }
+
+            match self.socket.poll_flush_unpin(cx)? {
+                Poll::Ready(()) => {}
+                Poll::Pending => {}
+            }
```

Comment:
I am wondering whether we should `flush` so early in the `poll` method, or whether it shouldn't be one of the last actions. Rational being that frequent flushing hurts performance, especially in case one can increase the batch instead.

Just a thought. Needs more thought and potentially data to back it up.

Reply:
Comment from reviewing commit-by-commit. This has been fixed in the meantime.

Reply:
I couldn't find any consistent performance improvement in my benchmarks when moving this block of code up or down.

However, this got me thinking: We do we communicate via channels between the connection and the stream for writing but we use shared buffers when reading? We could just as easily have a buffer of frames in `Shared` and wake the `Connection` whenever we write any frames to that. This would allow us to drain the buffer of all streams in one go, without having to receive individual frames over a channel.

Reply:
> I couldn't find any consistent performance improvement in my benchmarks when moving this block of code up or down.

Thanks for testing. Let's keep as is.

> However, this got me thinking: We do we communicate via channels between the connection and the stream for writing but we use shared buffers when reading? We could just as easily have a buffer of frames in `Shared` and wake the `Connection` whenever we write any frames to that. This would allow us to drain the buffer of all streams in one go, without having to receive individual frames over a channel.

I am decisive whether the connection should communicate with the stream via a channel or via plain `Mutex` and `Waker`. Whatever change we want to make, I think it should not happen within this pull request.